File: Lista 5/Python/Circulo.py
import math
from Ponto2D import Ponto2D

class Circulo:
    def __new__(cls, *args, **kwargs):
        return super().__new__(cls)

    def __init__(self, x = 0, y = 0, r = 1):
        if self.__validaCirculo(r):
            self.__x = x
            self.__y = y
            self.__r = r

    # Imprimir circulo
    def __str__(self):
        return f"({self.__x}, {self.__y}, {self.__r})"

    def __validaCirculo(self, r):
        if r > 0:
            return True
        raise ValueError("'r' deve ser maior que 0.")
    
    # Getters
    @property
    def x(self):
        return self.__x
    
    @property
    def y(self):
        return self.__y

    @property
    def r(self):
        return self.__r
    
    # Setters
    @x.setter
    def x(self, x):
        self.__x = x

    @y.setter
    def y(self, y):
        self.__y = y

    @r.setter
    def r(self, r):
        self.__r = r

    def isInnerPoint(self, ponto: Ponto2D):
        centro = Ponto2D(self.__x, self.__y)
        return ponto.distance(centro) <= self.__r

    def area(self):
        area = math.pi * self.__r**2
        return area

    def perimeter(self):
        perimeter = 2 * math.pi * self.__r
        return perimeter
    
    def isBiggerThan(self, circulo: "Circulo"):
        return self.area() > circulo.area()

if __name__ == "__main__":
    ponto = Ponto2D(-1, 2)
    circulo1 = Circulo(1, 3, 2)
    circulo2 = Circulo(r = 5)

    print(ponto)
    print(circulo2)
    print(f"Point {ponto} is inner of {circulo2}?", circulo2.isInnerPoint(ponto))
    print("Area: ", circulo1.area())
    print("Perimeter: ", circulo2.perimeter())
    print("Circulo1 is bigger than Circulo2?", circulo1.isBiggerThan(circulo2))

File: Lista 5/Java/Circulo.java
public class Circulo {
    private double x;
    private double y;
    private double r;

    public Circulo(double x, double y, double r) {
        validaCirculo(r);
        this.x = x;
        this.y = y;
        this.r = r;
    }

    private boolean validaCirculo(double r) {
        if (r > 0)
            return true;
        throw new IllegalArgumentException("Invalid r.");
    }

    // Setters
    public void setX(double x) {
        this.x = x;
    }

    public void setY(double y) {
        this.y = y;
    }

    public void setR(double r) {
        this.r = r;
    }

    // Getters
    public double getX() {
        return this.x;
    }
    
    public double getY() {
        return this.y;
    }

    public double getR() {
        return this.r;
    }

    public void imprimirCirculo() {
        System.out.println("(" + this.x + ", " + this.y + ", " + this.r + ")");
    }

    public boolean isInnerPoint (Ponto2D ponto) {
        Ponto2D centro = new Ponto2D(this.x, this.y);
        return ponto.distancia(centro) <= this.r;
    }

    public double area() {
        return Math.PI * this.r * this.r;
    }

    public double perimeter() {
        return 2 * Math.PI * this.r;
    }

    public boolean isBiggerThan(Circulo circulo) {
        return this.area() > circulo.area();
    }

    public static void main(String[] args) {
        Circulo circulo = new Circulo(0, 0, 3);
        Circulo circulo2 = new Circulo(0, 0, 1);
        circulo.imprimirCirculo();

        Ponto2D ponto = new Ponto2D(1 , 1);
        System.out.println(circulo.isInnerPoint(ponto));
        System.out.println(circulo.area());
        System.out.println(circulo.perimeter());
        System.out.println(circulo.isBiggerThan(circulo2));
    }

}


